domain/src/main/java/org/oppia/domain/topic/TopicController.kt

```diff
@@ -186,6 +220,170 @@ class TopicController @Inject constructor(
+  }
+
+  fun getOngoingTopicList(profileId: ProfileId): LiveData<AsyncResult<OngoingTopicList>> {
+    return dataProviders.convertToLiveData(retrieveOngoingTopicListDataProvider(profileId))
+  }
+
+  private fun retrieveCompletedStoryListDataProvider(profileId: ProfileId): DataProvider<CompletedStoryList> {
+    return dataProviders.transformAsync(
+      TRANSFORMED_GET_COMPLETED_STORIES_PROVIDER_ID,
+      storyProgressController.retrieveCacheStore(profileId)
+    ) {
+      val completedStoryListBuilder = CompletedStoryList.newBuilder()
+      it.topicProgressMap.keys.forEach { topicId ->
+        val topicProgress = it.topicProgressMap[topicId]
+        topicProgress!!.storyProgressMap.keys.forEach { storyId ->
+          val storyProgress = topicProgress.storyProgressMap[storyId]!!
+          val storySummary = retrieveStory(storyId)
+          val lastChapterSummary = storySummary.chapterList.last()
+          if (storyProgress.chapterProgressMap.containsKey(lastChapterSummary.explorationId)
+            && storyProgress.chapterProgressMap[lastChapterSummary.explorationId] == ChapterPlayState.COMPLETED
+          ) {
+            completedStoryListBuilder.addStorySummary(storySummary)
+          }
+        }
+      }
+      AsyncResult.success(completedStoryListBuilder.build())
+    }
+  }
+
+  private fun retrieveOngoingTopicListDataProvider(profileId: ProfileId): DataProvider<OngoingTopicList> {
+    return dataProviders.transformAsync(
+      TRANSFORMED_GET_ONGOING_TOPICS_PROVIDER_ID,
+      storyProgressController.retrieveCacheStore(profileId)
+    ) {
+      val ongoingTopicList = OngoingTopicList.newBuilder()
+      it.topicProgressMap.keys.forEach { topicId ->
+        val topic = retrieveTopic(topicId)
+        val topicProgress = it.topicProgressMap[topicId]
+        topic.storyList.forEach { storySummary ->
+          if (topicProgress!!.storyProgressMap.containsKey(storySummary.storyId)) {
+            val storyProgress = topicProgress.storyProgressMap[storySummary.storyId]
+            val lastChapterSummary = storySummary.chapterList.last()
+            if (!storyProgress!!.chapterProgressMap.containsKey(lastChapterSummary.explorationId)) {
+              ongoingTopicList.addTopic(topic)
+            }
+          }
+        }
+      }
+      AsyncResult.success(ongoingTopicList.build())
+    }
+  }
+
+  fun retrieveQuestionsForSkillIds(skillIdsList: List<String>): DataProvider<List<Question>> {
+    return dataProviders.createInMemoryDataProvider(QUESTION_DATA_PROVIDER_ID) {
+      loadQuestionsForSkillIds(skillIdsList)
+    }
+  }
+
+  /**
+   * Combines the specified topic and topic-progress into a new topic.
+   */
+  private fun combineTopicAndTopicProgress(topic: Topic, topicProgress: TopicProgress): Topic {
+    val topicBuilder = topic.toBuilder()
+    if (topicProgress.storyProgressMap.isNotEmpty()) {
+      topic.storyList.forEachIndexed { storyIndex, storySummary ->
```

Comment:
Can you explain a bit here & in a comment why the indexes here & below are important? They struck me as a bit odd because I'd normally expect that simply iterating across the collection would be sufficient.

Ditto for the other combine methods below.

Reply:
I was using the wrong function: `setStory` need index but `addStory` does not need any index.

Updated this
`topicBuilder.setStory(storyIndex, storyBuilder.build())`

to
`topicBuilder.addStory(storySummary)`

Reply:
I still see setStory, and indexed-based iteration. Was the update pushed?